simulator/memory/argv_loader/argv_loader.cpp

```diff
@@ -0,0 +1,82 @@
+/**
+ * argv_loader.cpp: loader of arguments to guest memory
+ * @author Rustem Yunusov <yunusov.rr@phystech.edu>
+ * Copyright 2019 MIPT-MIPS iLab project
+ */
+
+#include "argv_loader.h"
+
+#include <cstring>
+#include <infra/exception.h>
+
+template <typename T, Endian endian>
+ArgvLoader<T, endian>::ArgvLoader( const char* const* argv, const char* const* envp)
+        : argc( argv ? count_argc( argv) : 0)
+        , argv( argv)
+        , envp( envp)
+        , offset( 0)
+{}
+
+template <typename T, Endian endian>
+size_t ArgvLoader<T, endian>::load_to( const std::shared_ptr<FuncMemory>& mem, Addr addr)
+{
+    mem->write<T, endian>( narrow_cast<T>( argc), addr + offset);
+    offset += GUEST_WORD_SIZE;
+
+    offset += argc * bytewidth<Addr>; //reserved space for argv[]
+
+    place_nullptr( mem, addr + offset);
+    offset += bytewidth<Addr>;
+
+    if ( envp)
+    {
+        offset += count_argc( envp) * bytewidth<Addr>;
+
+        place_nullptr( mem, addr + offset);
+        offset += bytewidth<Addr>;
+    }
+
+    load_argv_contents( mem, addr);
+
+    if ( envp)
+        load_envp_contents( mem, addr);
+
+    return offset;
+}
+
+template<typename T, Endian endian>
+void ArgvLoader<T, endian>::load_argv_contents( const std::shared_ptr<FuncMemory>& mem, Addr addr)
+{
+    for ( int contents_offset = 0; contents_offset < argc; contents_offset++)
+    {
+        mem->write<Addr, endian>( addr + offset, addr + GUEST_WORD_SIZE + contents_offset * bytewidth<Addr>);
```

Comment:
Why did you revert this?
We always store T to guest memory, and we always use Addr to address memory.
Consider `Addr` as a guest equivalent of `int*`, and `T` as a guest equivalent of `int`.
The only "problematic" case occurs when we store Addr (which is 64 bit) to 32-bit machine memory. But in that case upper bits of Addr must be zero, so simple `narrow_cast` would solve all problems.

Reply:
As I understood:
firstly, we store T argc;
then we store Addr's of argv_contents;
then we store Addr's of envp_contents;
then we store chars, which are argv_contents;
lastly, we store chars, which are envp_contents.

Reply:
That's why I add bytewidth<Addr> to offset, not GUEST_WORD_SIZE.

Reply:
Now think how would 32-bit guest operating system read it:
`32 bits` of argc
`64 bits * argc` of argv conents